diag_manager/fms_diag_outfield.F90

```diff
@@ -0,0 +1,212 @@
+
+
+   !> @brief Class fms_diag_outfield_index_type which (along with class fms_diag_outfield_type)
+   !! encapsulate related information used in updating the output buffers by the diag_manager
+   !! send_data routines. This class in particular focuses on do loop index controls or settings.
+   !! Note that the index names in this class should be indentical to the names used in the
+   !! diag_manager send_data functions and in the "math" buffer update functions. The purpose
+   !! of this class is also to allow for a smaller call function signature for the math/buffer
+   !! update functions.
+   !> @ingroup fms_diag_outfield_mod
+   TYPE, public :: fms_diag_outfield_index_type
+      INTEGER :: f1,f2 !< Indecies used specify 1st dim bounds of field, mask and rmask.
+      INTEGER :: f3,f4 !< Indecies used specify 2st dim bounds of field, mask and rmask.
+      INTEGER :: is, js, ks  !< Start indecies in each spatial dim of the field_data; and
+      !! may be user provided in send_data
+      Integer :: ie, je, ke  !< End indecies in each spatial dim of the field_data; and
+      !! may be user provided in send_data
+      INTEGER :: hi !< halo size in x direction. Same name as in send_data
+      INTEGER :: hj !< halo size in y direction. Same
+   CONTAINS
+      procedure :: initialize => initialize_outfield_index_type
+   END TYPE fms_diag_outfield_index_type
+
+CONTAINS
+   !!TODO: In the modern diag, the field_val and weight may also be of integer type,
+   !!      and so may need to use the pre-processor.
+
+
+!> #brief initialize all the memebers of the class.
+   SUBROUTINE initialize_outfield_index_type(this, is, js , ks, ie, je, ke, hi, hj, f1, f2, f3, f4)
+      CLASS(fms_diag_outfield_index_type), INTENT(inout)  :: this
+      INTEGER, INTENT(in) :: is, js, ks !< Variable used to update class member of same names.
+      INTEGER, INTENT(in) :: ie, je, ke !< Variable used to update class member of same names.
+      INTEGER, INTENT(in) :: hi, hj !< Variable used to update class member of same names.
+      INTEGER, INTENT(in) :: f1, f2, f3, f4 !< Variable used to update class member of same names.
+
+      this%is = is
+      this%js = js
+      this%ks = ks
+      this%ie = ie
+      this%je = je
+      this%ke = ke
+
+      this%hi = hi
+      this%hj = hj
+
+      this%f1 = f1
+      this%f2 = f2
+      this%f3 = f3
+      this%f4 = f4
+   END SUBROUTINE initialize_outfield_index_type
+
+
+   !> @brief Update with those fields used in the legacy diag manager.
+  !! Note that this is initializing from the legacy structures.
+   !! Note that output_frequency  came from file_type;
+   SUBROUTINE initialize_outfield_imp(this, input_field,  output_field, mask_present, freq)
+      CLASS(fms_diag_outfield_type), INTENT(inout) :: this
+      TYPE(input_field_type),     INTENT(in) :: input_field
+      TYPE(output_field_type),    INTENT(in) :: output_field
+      LOGICAL, INTENT(in) :: mask_present
+      INTEGER, INTENT(in) :: freq
```

Comment:
These variables need to be documented in doxygen format

Reply:
done